be/src/storage/vectorized/schema_change.h

```diff
@@ -0,0 +1,223 @@
+    bool is_memory_enough_to_sort(size_t num_rows, size_t allocated_rows);
+
+private:
+    const TabletSchema& _tablet_schema;
+    size_t _memory_allocated;
+    size_t _row_len;
+    size_t _memory_limitation;
+};
+
+class SchemaChange {
+public:
+    SchemaChange(MemTracker* mem_tracker) : _filtered_rows(0), _merged_rows(0) {
+        _mem_tracker = std::make_unique<MemTracker>(-1, "", mem_tracker, true);
+    }
+    virtual ~SchemaChange() {}
+
+    virtual bool process(vectorized::Reader* reader, RowsetWriter* new_rowset_writer, TabletSharedPtr tablet,
+                         TabletSharedPtr base_tablet, RowsetSharedPtr rowset) = 0;
+
+    void add_filtered_rows(uint64_t filtered_rows) { _filtered_rows += filtered_rows; }
+
+    void add_merged_rows(uint64_t merged_rows) { _merged_rows += merged_rows; }
+
+    uint64_t filtered_rows() const { return _filtered_rows; }
+
+    uint64_t merged_rows() const { return _merged_rows; }
+
+    void reset_filtered_rows() { _filtered_rows = 0; }
+
+    void reset_merged_rows() { _merged_rows = 0; }
+
+protected:
+    std::unique_ptr<MemTracker> _mem_tracker = nullptr;
+
+private:
+    uint64_t _filtered_rows;
+    uint64_t _merged_rows;
+};
+
+class LinkedSchemaChange : public SchemaChange {
+public:
+    explicit LinkedSchemaChange(MemTracker* mem_tracker, ChunkChanger& chunk_changer)
+            : SchemaChange(mem_tracker), _chunk_changer(chunk_changer) {}
+    ~LinkedSchemaChange() { _mem_tracker->release(_mem_tracker->consumption()); }
+
+    bool process(vectorized::Reader* reader, RowsetWriter* new_rowset_writer, TabletSharedPtr new_tablet,
+                 TabletSharedPtr base_tablet, RowsetSharedPtr rowset) override;
+
+private:
+    ChunkChanger& _chunk_changer;
+    DISALLOW_COPY_AND_ASSIGN(LinkedSchemaChange);
+};
+
+// @brief schema change without sorting.
+class SchemaChangeDirectly : public SchemaChange {
+public:
+    explicit SchemaChangeDirectly(MemTracker* mem_tracker, ChunkChanger& chunk_changer)
+            : SchemaChange(mem_tracker), _chunk_changer(chunk_changer), _chunk_allocator(nullptr) {}
+    virtual ~SchemaChangeDirectly() {
+        SAFE_DELETE(_chunk_allocator);
+        _mem_tracker->release(_mem_tracker->consumption());
+    }
+
+    virtual bool process(vectorized::Reader* reader, RowsetWriter* new_rowset_writer, TabletSharedPtr new_tablet,
+                         TabletSharedPtr base_tablet, RowsetSharedPtr rowset) override;
+
+private:
+    ChunkChanger& _chunk_changer;
+    ChunkAllocator* _chunk_allocator;
+    DISALLOW_COPY_AND_ASSIGN(SchemaChangeDirectly);
+};
+
+// @breif schema change with sorting
+class SchemaChangeWithSorting : public SchemaChange {
+public:
+    explicit SchemaChangeWithSorting(MemTracker* mem_tracker, ChunkChanger& chunk_changer, size_t memory_limitation);
+    virtual ~SchemaChangeWithSorting();
+
+    virtual bool process(vectorized::Reader* reader, RowsetWriter* new_rowset_writer, TabletSharedPtr new_tablet,
```

Comment:
override is already used, virtual can be removed.

Reply:
done